### PR TITLE
Add logic to collect `VisLayer`s in line chart render flow

### LINKED ISSUE #3122: [Feature Anywhere] Collect `VisLayer`s when rendering visualization line charts
As part of #2880, we will need to inject logic in the visualization render flow to fetch and render any associated `VisLayer`s on the chart.
Breaking this into steps, this would look like:
1. Determine whether a given vis is eligible to render `VisLayer`s (e.g., of type `line`, valid axes configs)
2. Fetch all `augment-vis` saved objects that reference this particular vis
3. Construct a pipeline of expressions functions using the specified name & args listed in the saved object(s)
4. Execute the pipeline to collect a set of `VisLayer`s

The process for parsing the collected `VisLayer`s and actually updating the chart canvas is tracked in #3130.

### PR DESCRIPTION
**Pull request by ohltyler**
### Description
This PR adds the logic for collecting & parsing `VisLayer`s to eventually be read by the downstream render expressions functions. More details can be found in the [related issue](https://github.com/opensearch-project/OpenSearch-Dashboards/issues/3122)

Most of the new logic is happening in the `fetchVisLayers()` function of `visualize_embeddable` (it is calling new helper functions added in `utils`). This is ran whenever there are changes that affect the embeddable such that it needs to be re-rendered. In our use case, we will want to dynamically fetch updated `VisLayer`s whenever there are such changes, since the layers themselves are dynamically generated based on the context for the existing visualization (time range, filter, query). After any `VisLayer`s are collected, they are passed to the final `buildPipeline` fn which constructs the end-to-end string pipeline to execute and render the entire visualization.

The final processing of the `VisLayer`s will be handled in 2 more PRs:
1. Setting up the `toExpressionAst()` fn for line chart visualizations as part of #3106 
2. Add logic in the new expressions function to parse any `VisLayer`s and dynamically update the vega spec as part of #3145 

Other things to note:
1. There are some housekeeping changes where files from `vis_augmenter/common` are moved to `vis_augmenter/public`. This affects the files themselves + where they were being imported. **There is no changes to the content of the files, so they don't need much review/attention**
2. The `isEligibleForVisLayers()` helper fn is a placeholder for now - deeper logic will be added for that in the related issue  #3268. I've linked that in the code, and will also add a note in that issue to make sure this is tracked properly
 
### Issues Resolved
Closes #3122 
 
### Check List
- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
  - [ ] `yarn test:ftr`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [s] Commits are signed per the DCO using --signoff

### CODE REVIEW

**Comment by ohltyler** [comment]
Some of the new imports won't resolve until earlier PRs have been merged and I can rebase here. Until then you can ignore those `TODO` comments

**Comment by ohltyler** [comment]
PR is rebased and ready. **Please read PR description** before starting the review

File: src/plugins/visualizations/public/embeddable/visualize_embeddable.ts
Comment (amsiglan) [inline src/plugins/visualizations/public/embeddable/visualize_embeddable.ts:150]: Nit: We can declare this private here which will combine lines `138` and `173` below
`private savedAugmentVisLoader?: SavedAugmentVisLoader`

File: src/plugins/vis_augmenter/public/utils/utils.ts
Comment (amsiglan) [inline src/plugins/vis_augmenter/public/utils/utils.ts:44]: *Doubt:* We call this as expressionFn but I am not sure if it has an actual function implementation associated with it because from [the type](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/feature/feature-anywhere/src/plugins/vis_augmenter/public/types.ts#L19) I see it has a name and args associated with it. Can you point me to a place that can help me understand this better? Thanks!

**Review by amsiglan — APPROVED** [review]
Looks good to me overall.

File: src/plugins/visualizations/public/embeddable/visualize_embeddable.ts
Comment (ohltyler) [reply src/plugins/visualizations/public/embeddable/visualize_embeddable.ts:150]: Not sure what you mean here. It is declared private on 138, and line 150 is just specifying an arg in the constructor. I was following the patterns for the rest of the `visualize_embeddable` private fields, and am consistent with the existing `savedVisualizationsLoader`

File: src/plugins/vis_augmenter/public/utils/utils.ts
Comment (ohltyler) [reply src/plugins/vis_augmenter/public/utils/utils.ts:44]: Sure - it's defined in `src/plugins/vis_augmenter/public/types.ts` in #3109. Essentially it's an interface that defines how to run a plugin-defined expression function. `type` represents the return type of the expression function (same as existing expression functions), `name` is the name of the registered function, and `args` is the loosely-typed field for plugins to add their own args. An example would be something like an `add_anomalies` function defined in the AD plugin, which would collect anomalies and return the return type.

These functions are then read here and parsed out to create a single expressions string pipeline, where when executed, will collect all `VisLayer`s.

See details and diagrams in the meta issue #2880

**Comment by abbyhu2000** [comment]
@ohltyler I re-ran the tests couple times, seem like it is failing some of the vis builder functional tests. 
```
1)    visBuilder app
     │       Basic tests for visBuilder app 
     │         should show visualization when a field is added:
     │
     │      StaleElementReferenceError: stale element reference: element is not attached to the page document
     │   (Session info: headless chrome=110.0.5481.77)
  
     │2)    visBuilder app
     │       Basic tests for visBuilder app 
     │         should clear visualization when field is deleted:
     │
     │      Error: retry.try timeout: TimeoutError: Waiting for element to be located By(css selector, [data-test-subj="dropBoxField-metric-0"] [data-test-subj="dropBoxRemoveBtn"])
```

**Comment by ohltyler** [comment]
> @ohltyler I re-ran the tests couple times, seem like it is failing some of the vis builder functional tests.
> 
> ```
> 1)    visBuilder app
>      │       Basic tests for visBuilder app 
>      │         should show visualization when a field is added:
>      │
>      │      StaleElementReferenceError: stale element reference: element is not attached to the page document
>      │   (Session info: headless chrome=110.0.5481.77)
>   
>      │2)    visBuilder app
>      │       Basic tests for visBuilder app 
>      │         should clear visualization when field is deleted:
>      │
>      │      Error: retry.try timeout: TimeoutError: Waiting for element to be located By(css selector, [data-test-subj="dropBoxField-metric-0"] [data-test-subj="dropBoxRemoveBtn"])
> ```

this is unrelated to our change and could be due to the ongoing visbuilder changes and what is currently merged in our feature branch. I think this can be ignored and we can ensure all tests are passing consistently when finally merging into `main`/`2.x`. 

cc @joshuarrrr who is tracking any errors that pop up in these runs

**Review by abbyhu2000 — COMMENTED** [review]
LGTM! Just commented a couple nits.

File: src/plugins/vis_augmenter/public/types.ts
Comment (ohltyler) [reply src/plugins/vis_augmenter/public/types.ts:79]: `any` is for the arguments field, and it will need to stay loosely typed so plugins that utilize this function definition aren't restricted with how to pass arguments; we leave it up to the plugins to pass any arguments in any way they want.

File: src/plugins/visualizations/public/embeddable/visualize_embeddable.ts
Comment (ohltyler) [reply src/plugins/visualizations/public/embeddable/visualize_embeddable.ts:420]: Good catch - will update.

File: src/plugins/visualizations/public/embeddable/visualize_embeddable.ts
Comment (ohltyler) [reply src/plugins/visualizations/public/embeddable/visualize_embeddable.ts:509]: I prefer to leave as-is to follow existing functionality. Typically the types aren't declared as constants and imported.

File: src/plugins/vis_augmenter/public/utils/utils.ts
Comment (ohltyler) [reply src/plugins/vis_augmenter/public/utils/utils.ts:22]: Sure, I will add.

**Comment by ohltyler** [comment]
@abbyhu2000 I've refactored a few of the types & interfaces into separate `types.ts` files so they are more organized. The only change was making the `VisLayerFunctionDefinition` more typesafe by using the existing `ExprVisLayers` as the input/output.

File: src/plugins/vis_augmenter/public/types.ts
Comment (abbyhu2000) [reply src/plugins/vis_augmenter/public/types.ts]: make sense, thanks!

File: src/plugins/vis_augmenter/public/utils/utils.ts
Comment (abbyhu2000) [reply src/plugins/vis_augmenter/public/utils/utils.ts:27]: Thank you!

**Review by abbyhu2000 — APPROVED** [review]
@ohltyler Thank you for addressing the comments!

File: src/plugins/vis_augmenter/public/utils/utils.ts
Comment (joshuarrrr) [inline src/plugins/vis_augmenter/public/utils/utils.ts:7]: nit - for consistency

```suggestion
import { Vis } from '../../../../plugins/visualizations/public';
```

File: src/plugins/vis_augmenter/public/utils/utils.ts
Comment (joshuarrrr) [inline src/plugins/vis_augmenter/public/utils/utils.ts:34]: instead of using `findAll`, is there a more specific `find` method we can use (or define on SavedAugmentVisLoader) so that we can just fetch the ones we want?

File: src/plugins/vis_augmenter/public/utils/utils.ts
Comment (joshuarrrr) [inline src/plugins/vis_augmenter/public/utils/utils.ts:53]: nit - why `forEach` instead of just `map`?

**Review by joshuarrrr — APPROVED** [review]
Some nits and code style suggestions, but no blockers. I think using `find` instead of `findAll` would be the most significant improvement.

Also, thanks @ohltyler for the useful PR description!

File: src/plugins/visualizations/public/legacy/build_pipeline.ts
Comment (ohltyler) [reply src/plugins/visualizations/public/legacy/build_pipeline.ts:36]: sure i'll change. overall I've seen this not very consistent, is there any suggested styling on importing with `plugins/*` prefix?

File: src/plugins/visualizations/public/embeddable/visualize_embeddable.ts
Comment (ohltyler) [reply src/plugins/visualizations/public/embeddable/visualize_embeddable.ts:413]: Yeah that makes sense. I'll move that into the fn, & simplify the return type from `ExprVisLayers` => `VisLayers`.

File: src/plugins/visualizations/public/embeddable/visualize_embeddable.ts
Comment (ohltyler) [reply src/plugins/visualizations/public/embeddable/visualize_embeddable.ts:520]: I've simplified to have multiple `return` statements.

File: src/plugins/vis_augmenter/public/utils/utils.ts
Comment (ohltyler) [reply src/plugins/vis_augmenter/public/utils/utils.ts:34]: [findall](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/plugins/saved_objects/public/saved_object/saved_object_loader.ts#L131) and [find](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/plugins/saved_objects/public/saved_object/saved_object_loader.ts#L150) are almost identical, and both allow to pass some query values to narrow down the search. By default the `title` and `description` fields aren't helpful in our use case since we need to search by the vis ID field instead.

But, I do think we could override `findAll()` to include the vis ID field as an option that we can use in the query. Instead of blocking this PR and for the sake of time, can I open a follow-up issue to look into and potentially optimize this call?

File: src/plugins/vis_augmenter/public/utils/utils.ts
Comment (ohltyler) [reply src/plugins/vis_augmenter/public/utils/utils.ts:53]: Good call. Nothing else is going on in there so I will change to `map` and fill `visLayerExpressionFns` array automatically.

File: src/plugins/visualizations/public/embeddable/visualize_embeddable.ts
Comment (joshuarrrr) [inline src/plugins/visualizations/public/embeddable/visualize_embeddable.ts:413]: This will fix the test failure, I think.

```suggestion
      visLayers,
```

**Review by joshuarrrr — COMMENTED** [review]
one linting failure - you should be able to accept my change, and then I'll re-approve.

File: src/plugins/vis_augmenter/public/utils/utils.ts
Comment (ohltyler) [reply src/plugins/vis_augmenter/public/utils/utils.ts:34]: Created and linked in meta issue enhancement list: https://github.com/opensearch-project/OpenSearch-Dashboards/issues/3566

File: src/plugins/visualizations/public/embeddable/visualize_embeddable.ts
Comment (ohltyler) [reply src/plugins/visualizations/public/embeddable/visualize_embeddable.ts]: whoops! I had made that change when seeing it fail in the commit hook but forgot to add. Thanks

File: src/plugins/visualizations/public/embeddable/create_vis_embeddable_from_object.ts
Comment (amitgalitz) [inline src/plugins/visualizations/public/embeddable/create_vis_embeddable_from_object.ts:92]: If we are calling on getSavedAugmentVisLoader() inside the visualizations plugin don't we need to first set it in here: `src/plugins/visualizations/public/plugin.ts` @joshuarrrr

Comment (ohltyler) [reply src/plugins/visualizations/public/embeddable/create_vis_embeddable_from_object.ts:92]: Thanks, yes this was a miss. You've opened https://github.com/opensearch-project/OpenSearch-Dashboards/pull/3616 so I'll resolve here.

**Pull request merged**
